pkg/controller/infinispan/infinispan_controller.go

```diff
@@ -1743,18 +1716,67 @@ func appendKubernetesRemoteLocation(xsite *ispnutil.XSite, infinispan *infinispa
 	return nil
 }
 
-func getRemoteSiteServiceHostPort(service *corev1.Service, remoteKubernetes *ispnutil.Kubernetes, logger logr.Logger) (string, int32, error) {
+func getCrossSiteServiceHostPort(service *corev1.Service, kubernetes *ispnutil.Kubernetes, logger logr.Logger) (string, int32, error) {
 	switch serviceType := service.Spec.Type; serviceType {
 	case corev1.ServiceTypeNodePort:
 		// If configuring NodePort, expect external IPs to be configured
-		return remoteKubernetes.PublicIP(), remoteKubernetes.GetNodePort(service), nil
+		return getNodePortServiceHostPort(service, kubernetes, logger)
 	case corev1.ServiceTypeLoadBalancer:
 		return getLoadBalancerServiceHostPort(service, logger)
 	default:
 		return "", 0, fmt.Errorf("unsupported service type '%v'", serviceType)
 	}
 }
 
+func getNodePortServiceHostPort(service *corev1.Service, k *ispnutil.Kubernetes, logger logr.Logger) (string, int32, error) {
+	//The NodePort can be found in the service
+	port := service.Spec.Ports[0].NodePort
+
+	//The IPs must be fetch. Some cases, the API server (which handles REST requests) isn't the same as the worker
+	//So, we get the workers list. It needs some permissions cluster-reader permission
+	//oc create clusterrolebinding <name> -n ${NAMESPACE} --clusterrole=cluster-reader --serviceaccount=${NAMESPACE}:<account-name>
+	workerList := &corev1.NodeList{}
+
+	//select workers only. AFAIK, only the workers have the proxy to the pods
+	labelSelector := labels.SelectorFromSet(map[string]string{"node-role.kubernetes.io/worker": ""})
+
+	listOps := &client.ListOptions{
+		LabelSelector: labelSelector,
+	}
+	err := k.Client.List(context.TODO(), workerList, listOps)
+	if err != nil {
+		return "", service.Spec.Ports[0].NodePort, err
+	}
+
+	//host := k.PublicIP() //returns REST API endpoint. not good
+	//Pick one node "randomly" (first one)
+	host := workerList.Items[0].Status.Addresses[0].Address
+	logger.Info("Found NodePort endpoint", "Host", host, "Port", port)
+	return host, port, nil
+}
+
+func getLoadBalancerServiceHostPort(service *corev1.Service, logger logr.Logger) (string, int32, error) {
+	port := service.Spec.Ports[0].Port
+
+	// If configuring load balancer, look for external ingress
+	if len(service.Status.LoadBalancer.Ingress) > 0 {
+		ingress := service.Status.LoadBalancer.Ingress[0]
```

Comment:
@pruivo may be we need to check that Ingress[0] is present to prevent _Index out of range_?

Reply:
isn't `if len(service.Status.LoadBalancer.Ingress) > 0 {` enough?

Reply:
Oh, sorry, point on incorrect line